epick_description/urdf/epick_macro.xacro

```diff
@@ -0,0 +1,95 @@
+               params="prefix
+                       parent
+                       *origin
+                       extension_radius:=0.0
+                       extension_length:=0.0
+                       suction_cup_radius:=0.012
+                       suction_cup_height:=0.015">
+
+    <!-- Base joint -->
+    <joint name="${prefix}epick_base_joint" type="fixed">
+      <parent link="${parent}" />
+      <child link="${prefix}epick_body" />
+      <xacro:insert_block name="origin" />
+    </joint>
+
+
+    <!-- EPick body -->
+    <xacro:property name="body_length" value="0.1023" />
+    <link name="${prefix}epick_body">
+      <visual>
+        <origin xyz="0 0 0" rpy="0 0 0"/>
+        <geometry>          
+          <mesh filename="package://epick_description/meshes/visual/epick_body.stl"
+            scale="0.001 0.001 0.001"/>          
+        </geometry>
+        <material name="epick_gray">
+          <color rgba="0.25 0.25 0.25 1"/>
+        </material>        
+      </visual>
+      <collision>
+        <origin xyz="0 0 ${body_length / 2}" rpy="0 0 0"/>
+        <geometry>          
+          <cylinder radius="0.044" length="${body_length}"/>
+        </geometry>       
+      </collision>
+    </link>
+
+
+    <!-- Suction Cup (optional) --> 
+    <xacro:if value="${suction_cup_radius * suction_cup_height > 0.0}">
+      <link name="${prefix}epick_suction">
+
+        <!-- Extension (optional) -->
+        <xacro:if value="${extension_length * extension_radius > 0.0}">
+          <visual>
+            <origin xyz="0 0 ${extension_length / 2}" rpy="0 0 0"/>
+            <geometry>          
+              <cylinder radius="${extension_radius}" length="${extension_length}"/>
+            </geometry>
+            <material name="epick_gray"/>
+          </visual>
+        </xacro:if>
+
+        <!-- Suction Cup visual-->
+        <visual>
+          <origin xyz="0 0 ${extension_length + suction_cup_height / 2}" rpy="0 0 0"/>
+          <geometry>          
+            <cylinder radius="${suction_cup_radius}" length="${suction_cup_height}"/>
+          </geometry>
+          <material name="epick_gray"/>
+        </visual>
+
+        <!-- Collision Cylinder, includes both extension and cup -->
+        <collision>
+          <origin xyz="0 0 ${(extension_length + suction_cup_height) / 2}" rpy="0 0 0"/>
+          <geometry>          
+            <cylinder radius="${suction_cup_radius}" length="${extension_length + suction_cup_height}"/>
+          </geometry>
+          <material name="epick_gray"/>
+        </collision>
+      </link>
+
+      <!-- Connect suction cup to body -->
+      <joint name="${prefix}epick_body_to_suction" type="fixed">
+        <parent link="${prefix}epick_body"/>
+        <child link="${prefix}epick_suction"/>
+        <origin xyz="0 0 ${body_length}" rpy="0 0 0"/>
+      </joint>
+    </xacro:if>
+
+
+    <!-- Tip frame to use for IK, Z-axis pointing "forwards" -->
```

Comment:
This extra link could be removed if "epick_suction" was placed at the tip. Thoughts?

Reply:
Hmm... I think just having one frame at the suction tip is probably okay.

For compliant control / ensuring that the vacuum cups establish contact with an object, you have to deliberately plan a few cm in anyhow.

Reply:
good point! I'll add some option to configure this